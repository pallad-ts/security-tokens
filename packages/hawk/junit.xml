<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="jest tests" tests="2" failures="0" errors="0" time="10.25">
  <testsuite name="HAWKServerEngine" errors="0" failures="0" skipped="0" timestamp="2022-06-11T13:05:02" time="10.115" tests="2">
    <testcase classname="HAWKServerEngine verification with payload simple" name="HAWKServerEngine verification with payload simple" time="4.058">
    </testcase>
    <testcase classname="HAWKServerEngine verification with payload simple that fails" name="HAWKServerEngine verification with payload simple that fails" time="4.011">
    </testcase>
  </testsuite>
</testsuites>