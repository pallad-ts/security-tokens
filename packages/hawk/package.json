{
  "name": "@pallad/security-tokens-hawk",
  "version": "1.4.0",
  "description": "",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "files": [
    "compiled"
  ],
  "directories": {
    "test": "test"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "npx pallad-scripts test",
    "lint": "npx pallad-scripts lint",
    "generate-barrels": "npx barrelsby -l all --delete -d ./src",
    "compile": "npm run generate-barrels && npx pallad-scripts compile",
    "prepublishOnly": "npx pallad-scripts compile"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl)",
  "license": "MIT",
  "dependencies": {
    "@pallad/errors": "^1",
    "@pallad/secret": "^1.1.1",
    "@pallad/security-tokens": "^1.2.0",
    "@pallad/security-tokens-http": "^1.1.1",
    "alpha-errors": "^0.4.0",
    "hawk": "^9.0.1",
    "raw-body": "^2.5.1"
  },
  "devDependencies": {
    "@sweet-monads/either": "^3.1.0",
    "body-parser": "^1.20.0"
  },
  "gitHead": "344460311d3c794e74cf25c05808e68ddbdd92b4"
}
