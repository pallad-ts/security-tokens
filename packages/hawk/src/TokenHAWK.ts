import {Token} from '@pallad/security-tokens';

export class TokenHAWK extends Token {
	constructor(readonly user: string) {
		super();
	}
}

