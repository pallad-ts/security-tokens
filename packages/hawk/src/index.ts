export * from './CredentialsStorage';
export * from './errors';
export * from './HAWKServerEngine';
export * from './hawkVerification';
export * from './TokenHAWK';
