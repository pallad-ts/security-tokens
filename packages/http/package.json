{
  "name": "@pallad/security-tokens-http",
  "version": "1.1.0",
  "description": "",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "files": [
    "compiled"
  ],
  "directories": {
    "test": "test"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "../../node_modules/.bin/pallad-scripts test",
    "lint": "../../node_modules/.bin/pallad-scripts lint",
    "compile": "../../node_modules/.bin/pallad-scripts compile",
    "prepack": "../../node_modules/.bin/pallad-scripts compile"
  },
  "peerDependencies": {
    "@pallad/security-tokens": "*"
  },
  "devDependencies": {
    "@pallad/security-tokens": "^1.1.0"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl)",
  "license": "MIT"
}
