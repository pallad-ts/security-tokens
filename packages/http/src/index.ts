export * from './TokenFactory';
export * from './bearerHeader';
export * from './cookie';
export * from './header';