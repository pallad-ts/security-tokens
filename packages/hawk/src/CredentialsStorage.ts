import {Secret} from "@pallad/secret";
import {server} from 'hawk';
import {webcrypto} from 'node:crypto';
import {errors} from "./errors";

export class CredentialsStorage {
	private credentials = new Map<string, CredentialsStorage.Credential>();
	private usernameIndex = new Set<string>();
	private keyIndex = new Set<string>();

	constructor(private algorithm: 'sha1' | 'sha256') {
	}

	registerCredential(id: string, key: Secret<string>, user: string) {
		if (this.usernameIndex.has(user)) {
			throw errors.USER_ALREADY_ALREADY_REGISTERED.format(user);
		}

		if (this.keyIndex.has(id)) {
			throw errors.CREDENTIAL_WITH_ID_ALREADY_REGISTERED.format(id);
		}

		this.credentials.set(id, {
			key,
			user
		});
		this.usernameIndex.add(user);

		return this;
	}

	retrieveCredentials(id: string): server.Credentials {
		if (!this.credentials.has(id)) {
			throw errors.NO_CREDENTIALS_FOUND.format(id);
		}
		const credential = this.credentials.get(id)!;

		return {key: credential.key.getValue(), user: credential.user, algorithm: this.algorithm};
	}
}

export namespace CredentialsStorage {
	export interface Credential {
		key: Secret<string>,
		user: string;
	}
}
