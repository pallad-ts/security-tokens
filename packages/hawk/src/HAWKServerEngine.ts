import { server } from "hawk";
import * as http from "http";

import { CredentialsLoader } from "./CredentialsLoader";
import { TokenHAWK } from "./TokenHAWK";
import { errors } from "./errors";

import getRawBody from "raw-body";

export class HAWKServerEngine {
	private finalCredentialsLoader = async (id: string) => {
		const credentials = await this.credentialsLoader(id);
		if (credentials === undefined) {
			throw errors.NO_CREDENTIALS_FOUND.create(id);
		}
		return credentials;
	};

	constructor(private credentialsLoader: CredentialsLoader) {}

	async verifyWithoutPayload(request: HAWKServerEngine.Request, options?: HAWKServerEngine.Options) {
		const result = await this.internalVerify(request, options);
		return new TokenHAWK(result.user);
	}

	private async internalVerify(
		request: http.IncomingMessage,
		options?: HAWKServerEngine.Options & { payload?: Buffer }
	) {
		const { payload, contentType, ...restOptions } = options || {};
		const { credentials, artifacts } = await server.authenticate(request, this.finalCredentialsLoader, restOptions);

		if (payload) {
			server.authenticatePayload(
				payload as any,
				credentials,
				artifacts,
				contentType ?? request.headers["content-type"]!
			);
		}

		return credentials;
	}

	async verifyWithPayload(request: HAWKServerEngine.Request, options?: HAWKServerEngine.Options.WithPayload) {
		const payload = options?.payload ?? (await this.getPayloadFromRequest(request));
		const result = await this.internalVerify(request, {
			...options,
			payload,
		});

		return new TokenHAWK(result.user);
	}

	private async getPayloadFromRequest(request: HAWKServerEngine.Request) {
		if (request.method === "GET" || request.method === "HEAD") {
			return Buffer.from("", "utf-8");
		}

		return getRawBody(request);
	}
}

export namespace HAWKServerEngine {
	export interface Options {
		hostHeaderName?: string;
		host?: string;
		port?: number;
		contentType?: string;
	}

	export namespace Options {
		export interface WithPayload extends Options {
			payload?: Buffer;
		}
	}

	export type Request = http.IncomingMessage;
}
