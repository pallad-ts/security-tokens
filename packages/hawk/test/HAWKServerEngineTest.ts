import * as http from "http";
import {json} from 'body-parser';
import * as sinon from 'sinon';
import {CredentialsStorage} from "@src/CredentialsStorage";
import {secret} from "@pallad/secret";
import {HAWKServerEngine} from "@src/HAWKServerEngine";
import {TokenHAWK} from "@src/TokenHAWK";
import {request} from 'undici';
import {client} from 'hawk';
import {Either} from "monet";


describe('HAWKServerEngine', () => {
	let server: http.Server;
	let handler: sinon.SinonStub;
	let engine: HAWKServerEngine;
	const URL = `http://localhost:10000/some-path`;

	const credentialsStorage = new CredentialsStorage('sha256');
	credentialsStorage.registerCredential('1', secret('k1'), 'u1');
	credentialsStorage.registerCredential('2', secret('k2'), 'u2');

	beforeEach(() => {
		handler = sinon.stub();
		server = http.createServer((req, res) => {
			json()(req, res, () => {
				handler(req, res);
			});
		});
		server.listen(10000);
		engine = new HAWKServerEngine(credentialsStorage);
	});

	afterEach((done) => {
		server.close(done);
	});

	describe('verification with payload', () => {
		it('simple', async () => {
			let result: Promise<TokenHAWK> | undefined;
			handler.callsFake(async (req, res) => {
				result = engine.verifyWithPayload(req);
				res.end();
			});

			const payload = Buffer.from('somepayloaddata', 'utf-8');

			const {header} = client.header(URL, 'POST', {
				contentType: 'application/octet-stream',
				payload: payload as any,
				credentials: {id: '1', ...credentialsStorage.retrieveCredentials('1')}
			});

			await request(URL, {
				method: 'POST',
				headers: {Authorization: header, 'content-type': 'application/octet-stream'},
				body: payload
			});

			const resolvedResult = await result;
			expect(resolvedResult)
				.toBeInstanceOf(TokenHAWK);

			expect(resolvedResult)
				.toHaveProperty('user', 'u1');
		});

		it('simple that fails', async () => {
			let result: Promise<Either<unknown, unknown>> | undefined;
			handler.callsFake(async (req, res) => {
				res.end();
				result = Either.fromPromise(engine.verifyWithPayload(req));
			});

			const {header} = client.header(URL, 'POST', {
				contentType: 'application/octet-stream',
				payload: Buffer.from('somepayloaddata', 'utf-8') as any,
				credentials: {id: '1', ...credentialsStorage.retrieveCredentials('1')}
			});

			await request(URL, {
				method: 'POST',
				headers: {Authorization: header, 'content-type': 'application/octet-stream'},
				body: Buffer.from('completely different payload')
			});

			const resolvedResult = await result;

			expect(resolvedResult)
				.toMatchSnapshot();
		});
	});
});
