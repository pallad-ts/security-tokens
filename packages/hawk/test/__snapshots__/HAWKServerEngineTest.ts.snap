// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HAWKServerEngine verification with payload simple that fails 1`] = `
Object {
  "@@type": "monet.js/Either",
  "isRightValue": false,
  "value": [Error: Bad payload hash],
}
`;
